File: go/soaparse.go
package soaparse

import (
	"strings"
)

type SOArecord struct {
	mname string

	rname string
	parsed_email string

	serial string

	refresh string
	retry string
	expire string
	minimum string
}

// return index of second to last occurrence of char in string
func SecondToLast(str string, char string) int {
	return strings.LastIndex(str[:strings.LastIndex(str, char)], char)
}

// replace char given by index with replacement string
func ReplaceCharAt(str, replacement string, index int) string {
	return str[:index] + replacement + str[index + len(replacement):]
}

func EmailFromSOA(soaEmailField string) string {
	var i int = SecondToLast(soaEmailField, ".")
	var email string = ReplaceCharAt(soaEmailField, "@", i)
	return email
}

func SOAparse (soaString string) SOArecord {
	var parts []string = strings.Split(soaString, " ")

	var soa SOArecord
	soa.mname = strings.TrimSuffix(parts[0], ".")

	soa.rname = strings.TrimSuffix(parts[1], ".")
	soa.parsed_email = EmailFromSOA(soa.rname)

	soa.serial = parts[2]

	soa.refresh = parts[3]
	soa.retry = parts[4]
	soa.expire = parts[5]
	soa.minimum = parts[6]

	return soa
}

File: soaparse.js
function soaParse(soaString) {
    // https://www.ripe.net/publications/docs/ripe-203
    var soaRecord = {
        mname: "",

        rname: "",
        parsed_email: "",

        serial: "",

        refresh: null,
        retry: null,
        expire: null,
        minimum: null,
    };

    var soaParts = soaString.split(" ");

    soaRecord.mname = removeLastChar(soaParts[0]);

    soaRecord.rname = removeLastChar(soaParts[1]);
    soaRecord.parsed_email = emailFromSOA(soaRecord.rname);

    soaRecord.serial = soaParts[2];

    soaRecord.refresh = parseInt(soaParts[3]);
    soaRecord.retry = parseInt(soaParts[4]);
    soaRecord.expire = parseInt(soaParts[5]);
    soaRecord.minimum = parseInt(soaParts[6]);

    return soaRecord;
}

function emailFromSOA(soaEmailString) {
    /* Index of the second to last ".", likely the char where @ email address is
    assumes the email hostname does not have a subdomain */
    var i = secondToLastInstance(soaEmailString, ".");
    var parsedEmail = replaceAt(soaEmailString, "@", i);
    return parsedEmail;
}

function removeLastChar(str) {
    return str.substr(0, str.length - 1);
}

// Index of the second to last instance of char
function secondToLastInstance(string, char) {
    return string.lastIndexOf(char, string.lastIndexOf(char) - 1);
}

// https://stackoverflow.com/questions/1431094/how-do-i-replace-a-character-at-a-particular-index-in-javascript?rq=1
function replaceAt(str, replacement, i) {
    return str.substr(0, i) + replacement
    + str.substr(i + replacement.length);
}
